src/Functions/parseDateTime.cpp

```diff
@@ -0,0 +1,1747 @@
+
+            if (!isString(arguments[0].type))
+                throw Exception(
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                    "Illegal type {} of first argument of function {} when arguments size is 1. Should be string",
+                    arguments[0].type->getName(),
+                    getName());
+
+            if (arguments.size() > 1 && !isString(arguments[1].type))
+                throw Exception(
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                    "Illegal type {} of second argument of function {} when arguments size is 1. Should be string",
+                    arguments[0].type->getName(),
+                    getName());
+
+            if (arguments.size() > 2 && !isString(arguments[2].type))
+                throw Exception(
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                    "Illegal type {} of third argument of function {} when arguments size is 1. Should be string",
+                    arguments[0].type->getName(),
+                    getName());
+
+            String time_zone_name = getTimeZone(arguments).second;
+            return std::make_shared<DataTypeDateTime>(time_zone_name);
+        }
+
+        ColumnPtr
+        executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & /*result_type*/, size_t input_rows_count) const override
+        {
+            const auto * col_str = checkAndGetColumn<ColumnString>(arguments[0].column.get());
+            if (!col_str)
+                throw Exception(
+                    ErrorCodes::ILLEGAL_COLUMN,
+                    "Illegal column {} of first ('str') argument of function {}. Must be string.",
+                    arguments[0].column->getName(),
+                    getName());
+
+            String format = getFormat(arguments);
+            const auto * time_zone = getTimeZone(arguments).first;
+
+            std::vector<Action> instructions;
+            parseFormat(format, instructions);
+
+            auto col_res = ColumnDateTime::create();
+            col_res->reserve(input_rows_count);
+            auto & data_res = col_res->getData();
+            Date date;
+            for (size_t i = 0; i < input_rows_count; ++i)
+            {
+                date.reset();
+
+                StringRef str_ref = col_str->getDataAt(i);
+                Pos cur = str_ref.data;
+                Pos end = str_ref.data + str_ref.size;
+                for (const auto & instruction : instructions)
+                {
+                    // std::cout << "instruction:" << instruction.toString() << std::endl;
+                    cur = instruction.perform(cur, end, date);
+                    // std::cout << "date:" << date.toString() << std::endl;
+                }
+
+                // Ensure all input was consumed.
+                if (cur < end)
+                    throw Exception(
+                        ErrorCodes::LOGICAL_ERROR,
```

Comment:
That's not a logical error, please see my other comment and proposal for a better error code.